subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/TaskExecutionTracker.kt

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.configurationcache
+
+import org.gradle.api.Task
+import org.gradle.api.execution.TaskActionListener
+import org.gradle.internal.operations.BuildOperationAncestryTracker
+import org.gradle.internal.operations.BuildOperationRef
+import org.gradle.internal.operations.CurrentBuildOperationRef
+import org.gradle.internal.operations.OperationIdentifier
+import org.gradle.internal.service.scopes.EventScope
+import org.gradle.internal.service.scopes.Scopes
+import java.util.Objects
+
+
+/**
+ * Tracks task execution. Can be used to find when an included build runs some forbidden action as part of the task execution, where it should be allowed.
+ */
+@EventScope(Scopes.BuildTree::class)
+interface TaskExecutionTracker {
+    fun isCurrentThreadExecutingTaskBuildOperation(): Boolean
+}
+
+
+class DefaultTaskExecutionTracker(
+    private val buildOperationAncestryTracker: BuildOperationAncestryTracker,
+) : TaskExecutionTracker, TaskActionListener {
+    private
+    val currentBuildOperationRef = CurrentBuildOperationRef.instance()
+
+    private
+    val currentRunningTaskOperations = HashSet<BuildOperationRef>()
+
+    override fun beforeActions(task: Task) {
+        currentBuildOperationRef.get()?.let { currentRunningTaskOperations.add(it) }
+    }
+
+    override fun afterActions(task: Task) {
+        currentBuildOperationRef.get()?.let { currentRunningTaskOperations.remove(it) }
+    }
+
+    override fun isCurrentThreadExecutingTaskBuildOperation(): Boolean {
+        val currentBuildOperationId = currentBuildOperationRef.id
+        val taskAncestorId = buildOperationAncestryTracker.findClosestMatchingAncestor(currentBuildOperationId, ::isOperationWithIdRunningTaskNow)
+        return taskAncestorId.isPresent
+    }
+
+    private
+    fun isOperationWithIdRunningTaskNow(id: OperationIdentifier): Boolean {
+        return currentRunningTaskOperations.find { taskOperationRef -> Objects.equals(id, taskOperationRef.id) } != null
```

Comment:
❓ Can't we just match against one of the Task execution build operation types here?

Reply:
Could you please elaborate on this a bit? I see two relevant types in the sources:
- `BuildOperationCategory` seems to have all necessary types but I'm not sure that it is consistently available, e.g. I'm seeing `NONE` in my tests
- `BuildOperationType` seems to be more reliable but it has ~50 implementations, so I'm not sure if it is feasible to check for all possibilities. The type isn't always straightforward if, for example, the worker API is involved (kotlin compiler daemon does this).

Reply:
I meant matching on the `BuildOperationType` looking for something like`ExecuteTaskActionBuildOperationType`, https://github.com/gradle/gradle/blob/c14637ed0a5ef90974e6bac19670ee94a9bed44f/subprojects/enterprise-operations/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteTaskActionBuildOperationType.java#L26

The rationale is that when the current thread is executing a task action, a build operation of type `ExecuteTaskActionBuildOperationType` should be reachable from the innermost build operation held by `CurrentBuildOperationRef`. I'm not sure whether that also covers the worker API usages.

Reply:
Oh, I think I see now. We can avoid listening to the tasks and tracking them, and instead just find an ancestor with details of the aforementioned type. Worker operations have some different type but the ancestry chain is still valid there.

Reply:
> We can avoid listening to the tasks and tracking them, and instead just find an ancestor with details of the aforementioned type

Exactly! I'm sorry I should have been more explicit in my original question.

Reply:
Following our discussion, it looks like TaskActionListener is not suitable because it doesn't include up-to-date checks and the likes, which are allowed to execute external processes in the non-build-logic builds. Something that covers these calls (e.g. `TaskListenerInternal`) is necessary there.

> just find an ancestor with details of the aforementioned type

I've looked into it but without much luck. `BuildOperationAncestryTracker` only stores `OperationIdentifier`s from which I don't know a way to obtain a "details" object to check type, so I have to store ids of build operations that run tasks somewhere. It makes sense switching to `BuildOperationListener` though to minimize hidden assumptions like "the `TaskListenerInternal.beforeExecute` runs in the task's build operation".